### ISSUE TITLE
Add support for "required" modules

### ISSUE THREAD

**Issue by rcernich** [jboss-container-images/concreate#30]
Currently, all modules listed are included in the image and used during the build.  There may be cases where a module requires another module, but it does not make sense for that module to include its dependency directly.  The best example of this is for layered images like the kieserver image, where its modules may depend on functionality being provided by the core EAP modules, but those modules are already included in the image and don't need to be include, their existence just needs to be verified during the build.

Another use case would be with a more complicated dependency chain, whereby an aggregate eap module might include or require a core java module and a subsequent layered image might contain modules that require something from one of the java modules.  The graph might look something like:

jdv includes eap requires java, with the jdv image.yaml including both eap and java, thus fulfilling the eap module's requirement on java.

**Comment by goldmann** [comment]
I'm not sure if I understand this issue. Following your example you mentioned that EAP modules are already included in the image, so my question is - why we should/want to list these modules again in the "KIE" image?

**Comment by rcernich** [comment]
The notion is that a module for kieserver may depend on functionality that is provide by an upstream module.  In this case, we don't want to reinstall the module, but we do want to verify that it exists when we build the image, so we can fail if it is not present.  A concreate (sp? ;)) example would be os-kieserver-s2i, which requires os-eap-s2i.

**Comment by dbecvarik** [comment]
We need to write proper design document for this. I have a lot of questions like:

If module is required its only checked if exists - so it means, artifacts are not processed, scripts are not executed, module descriptor is not merged?  
As we can assume we can locate the scripts to run them, we are unable to fetch artifact, merge descriptor in runtime during image build -> script will be probably unusable too.

**Comment by rcernich** [comment]
> If module is required its only checked if exists - so it means, artifacts are not processed, scripts are not executed, module descriptor is not merged?

Correct, although it may be installed during the same build, e.g. if it's listed in the install section of another installed module or image.yaml.

> As we can assume we can locate the scripts to run them, we are unable to fetch artifact, merge descriptor in runtime during image build -> script will be probably unusable too.

We shouldn't be running anything, only verifying that the module is present in the image.  If the module is present, then its scripts have already been executed, and its artifacts, labels, etc. have already been installed in the image.

From an implementation perspective, we might look at the following:
1. have a modules registry in the image that can be used for verifying existence
2. add something into the Dockerfile that runs a verification, e.g. verify modules x, y and z exist in the image's module registry.  ideally, we'd want to do this during generation, but if the modules are already included in the base image, we can only do this as part of the image build.  this might look like a ``CMD verify-prerequisites module.x module.y`` where modules x, y and z are listed as required, with module z being installed by the image.yaml, so we've verified the existence of module.z during the generation phase and only need to verify x and y during the build phase.
3. if a required module is missing the build fails, either during generation or during image build.

The key here is to allow for the definition of a dependency chain that can be verified during the build.  I think the kieserver example above should be used as a concrete example/test case.  The kieserver s2i module requires the eap s2i module, but it is already installed in the base image.  We just need to verify that the eap s2i module is present when we do the kieserver build.

A side effect of defining requirements is that we should probably ensure the required module is installed before the module requiring it is installed.  I don't believe this to be an issue at this time, but it's something we should be aware of (e.g. if the install scripts are doing something to modify the other module).

**Comment by dbecvarik** [comment]
> We shouldn't be running anything, only verifying that the module is present in the image. If the module is present, then its scripts have already been executed, and its artifacts, labels, etc. have already been installed in the image.

So we can do this for example in a way like defining label like 'concreate.used.modules' which will contain a list of modules installed (used to build) an image and just check if the module name is in the label in 'from' image.
If its not being executed/used it doesn't have to be in image at all, we can just use metadata we create.
Am I missing something?

**Comment by rcernich** [comment]
The only piece you're missing is that it might already be installed in the base image and is not installed in the build.  The kieserver is the best (only?) example of this at the moment, because it is built from the eap64-openshift image.  As per the example above, the only way to verify that something like os-eap-s2i module is installed is to check what modules are installed in the base image.  This really means that verification needs to happen during image build, since there's no way to know before hand (because it won't be in the metadata used to generate the build file).

**Comment by rcernich** [comment]
Perhaps we emit a warning during generation, if a required module is missing, but don't error until the build.

**Comment by dbecvarik** [comment]
I meant it differently. We will create label in image itself​| which will
contain name of all installed modules. So we can inspect base image label
if module was installed there. So this will be visible via docker inspect.

**Comment by rcernich** [comment]
That's fine.  At one point, we had talked about including the module source in the images.  However, a label would certainly make things easer to view without having to crack into the image.

**Comment by dbecvarik** [comment]
Yes we can talk about puting sources inside images later. There is even
issue for this.
